### DISCUSSION TITLE
能否总结下Plonk或其他零知识系统中常见性能优化技巧？

### DISCUSSION THREAD

**Discussion by a7853z** [zkp-co-learning/zkp-co-learn#94] [Q&A（提问题在隔壁~）]
如题，能否简单总结下Plonk或其他零知识系统用到的性能优化技巧？哪些是通用的，值得我们在零知识实践中好好学习掌握？
比如目前看到的Plonk用到的有：

1. KZG证明聚合
2. FFT
3. t多项式降次减少srs长度
4. 多项式线性化

**Comment by Howard-Hu** [comment]
通常来说，性能优化是相对于一个现有的项目来说，以现有的性能指标为优化的初始起点。
可以分为理论和工程实现两个方面：理论上期望减少其计算量；工程上如可并行的进行并性，优化系统调度，资源的分配等等。

零知识证明系统的层次为：
- 有限域运算：即大数的模加模减模乘模逆等；
- 椭圆曲线运算/配对运算;
- FFT/MSM;
- 然后之上还有一些构成零知识证明系统；

下面我例举一些使用到KZG多项式的零知识证明系统的可能优化：
1. 在有限域运算这一层：看它的实现是否是高效的实现，如是否使用蒙哥马利模乘等；具体可参考https://cacr.uwaterloo.ca/hac/about/chap14.pdf
2. 椭圆曲线运算的实现：点加及倍点本身的实现，可参考https://www.hyperelliptic.org/EFD/g1p/auto-shortw-jacobian-0.html  ；另外也可进一步使用glv方法优化倍点运算的实现，可参考https://hackmd.io/@chaosma/SyAvcYFxh#GLV
3. 配对运算：配对运算一些实现中由miller_loop+final_exponentiation构成；其中final_exponentiatin部分也有一些文章可能可以参考：如https://eprint.iacr.org/2016/130.pdf. 和https://eprint.iacr.org/2020/875
4.FFT:FFT的实现可以使用cuda实现来实现并行加速；
5.MSM:使用了多项式承诺的方案的零知识系统都跳不开需要计算MSM，可以使用pippenger算法等提高性能；
这里给出pippenger算法论文以供参考：https://jbootle.github.io/Misc/pippenger.pdf
另外也可参考：https://hackmd.io/@drouyang/msm    给出的一些方式，如使用Batch Addition方式来减少求逆次数(有限域上的运算模逆运算开销是最大的)，具体看文章这里不再展开；

这些应该是基本符合通用性；
更多的细节或者再更上一层的希望接触更多的大佬来补全；
希望这个回答对你有一些帮助。

**Reply by a7853z** [reply]
多谢回复 上述优化技巧偏底层，大佬们继续补充！
